compiler-lib/benches/symtab_bench.rs

```diff
@@ -0,0 +1,27 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+
+use compiler_lib::symtab::Scoped;
+
+fn scope_walk(input: &'_ usize) {
+    let defs: Vec<String> = (0..*input).map(|i| format!("scope#{}", i)).collect();
+    let mut scoped = Scoped::new();
+    for def in defs.iter() {
+        scoped.define(def, ()).unwrap();
+        scoped.enter_scope();
+    }
+    for depth in (0..*input).rev() {
+        let def = &defs[depth];
```

Comment:
we should lookup `defs[0]`, not `defs[depth]`. this only the subsequent `.visible_definition()` should also be checked. I think it's returning None except for the first iteration of the loop.

Reply:
I'll improve this microbenchmark a little, and add another one that works on an actual AST.

Reply:
With `defs[0]`and the construction of `scoped` outside of the benchmark, we get positive time per op:

```
cope_walk/1            time:   [58.219 ns 58.449 ns 58.705 ns]                         
                        change: [-7.4808% -6.8786% -6.3780%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe
scope_walk/10           time:   [594.27 ns 602.37 ns 612.59 ns]                           
                        change: [-81.014% -80.831% -80.613%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe
scope_walk/50           time:   [2.9521 us 2.9720 us 2.9941 us]                           
                        change: [-95.784% -95.751% -95.719%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
scope_walk/100          time:   [5.9262 us 5.9879 us 6.0551 us]                            
                        change: [-98.002% -97.952% -97.905%] (p = 0.00 < 0.05)
                        Performance has improved.
scope_walk/500          time:   [29.092 us 29.215 us 29.359 us]                            
                        change: [-99.590% -99.586% -99.582%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe
```

Reply:
OK, won't write actual AST-climbing microbenchmark, too much work.
But given above results,  I'll add the HashMap based lookups onto the target branch.